base/broadcast.jl

```diff
@@ -969,15 +969,34 @@ end
     end
 end
 
-# For broadcasted assignments like `broadcast!(f, A, ..., A, ...)`, where `A`
-# appears on both the LHS and the RHS of the `.=`, then we know we're only
-# going to make one pass through the array, and even though `A` is aliasing
-# against itself, the mutations won't affect the result as the indices on the
-# LHS and RHS will always match. This is not true in general, but with the `.op=`
-# syntax it's fairly common for an argument to be `===` a source.
-broadcast_unalias(dest, src) = dest === src ? src : unalias(dest, src)
+"""
+    broadcast_unalias(dest, src)
+
+`broadcast_unalias` is a simple extension of `Base.unalias` to enable an important
+performance optimization specific to the semantics of broadcasting.
+
+For broadcasted assignments like `broadcast!(f, A, ..., A, ...)`, where `A`
+appears on both the LHS and the RHS of the `.=`, then we know we're only
+going to make one pass through the array, and even though `A` is aliasing
+against itself, the mutations won't affect the result as the indices on the
+LHS and RHS will always match. As long as the indices do not repeat, the
+RHS does not need to be unaliased. This is particuclarly valuable with the `.op=`
+syntax since that makes it fairly common for an argument to be `===` a source.
+"""
+broadcast_unalias(dest, src) = dest === src && !potentially_self_aliased(dest) ? src : unalias(dest, src)
 broadcast_unalias(::Nothing, src) = src
 
+"""
+    potentially_self_aliased(A)
```

Comment:
A year later, I'm not a big fan of this name.  I'm guessing I used `potentially` as an attempt to say that `true`s are conclusive but `false`s are not... but I don't think it actually reads that way.

`mightalias` documents itself as a "conservative" test for sharing the same memory, but I don't even know what _that_ means on its face.  Is it more likely to have false positive or a false negative?

In both cases, I used a different word than my usual go-to "maybe_foo" as an attempt to do _some_ work in one direction or the other, but I don't think it was successful because I'm having a hard time piecing together what they mean right now.  In short, here's what we have between these two systems:

| function | returns | wrong if... |
| - | - | - |
| `mightalias` | `true` | the arrays reference distinct subsets of the same memory region |
| `mightalias`  | `false` | a custom array has not tracked its internal field(s) with `dataids` and one of those aliases |
| `potentially_self_aliased` | `true` | an array goofed its implementation (the default definitions should not have false positives) |
| `potentially_self_aliased` | `false` | this is the default answer; a self-aliasing array didn't define a method on this obscure function |

Reply:
Numpy calls this `may_have_internal_overlap` with tri-state output (known to not overlap, known to overlap, and unknown).